python/src/iceberg/transforms.py

```diff
@@ -301,6 +301,86 @@ def _(self, value_type: IcebergType, value: int) -> str:
         return datetime.to_human_timestamptz(value)
 
 
+class TruncateTransform(Transform[S, S]):
+    """A transform for truncating a value to a specified width.
+    Args:
+      source_type (Type): An Iceberg Type of IntegerType, LongType, StringType, BinaryType or DecimalType
+      width (int): The truncate width
+    Raises:
+      ValueError: If a type is provided that is incompatible with a Truncate transform
+    """
+
+    def __init__(self, source_type: IcebergType, width: int):
+        super().__init__(
+            f"truncate[{width}]",
+            f"transforms.truncate(source_type={repr(source_type)}, width={width})",
+        )
+        self._type = source_type
+        self._width = width
+
+    @property
+    def width(self):
+        return self._width
+
+    def apply(self, value: Optional[S]) -> Optional[S]:
+        return self._truncate_value(value) if value is not None else None
+
+    @singledispatchmethod
+    def _truncate_value(self, value: S) -> S:
+        raise ValueError(f"Cannot truncate value: {value}")
+
+    @_truncate_value.register(int)
+    def _(self, value):
+        """Truncate a given int value into a given width if feasible."""
+        if type(self._type) in {IntegerType, LongType}:
```

Comment:
Having the validation in processing itself feels a bit weird to me, shouldn't we check this when initializing the transform?

Reply:
The validation here is to catch the case that the caller passes an int into Non-IntegerType, e.g. StringType truncate transformer. But the initialization might not the value type.

Reply:
I don't think that we should do this check. We can assume that the type matches because of how we bind expressions.

Reply:
+1 to not validating inside of the right loop.

I’d like to see general validation of the `source_type` (though I guess singledispatch takes care of that somewhat, it would be nice to see it a bit more explicitly but that’s just my opinion),

Also validating in the constructor or elsewhere that the `width` is greater than zero.

But this function will be called in a tight loop and it’s best to avoid expensive checks here.